rest/route/middleware.go

```diff
@@ -914,6 +917,37 @@ func (m *EventLogPermissionsMiddleware) ServeHTTP(rw http.ResponseWriter, r *htt
 	next(rw, r)
 }
 
+// NewGithubAuthMiddleware returns a middleware that verifies the payload
+func NewGithubAuthMiddleware() gimlet.Middleware {
+	return &githubAuthMiddleware{}
+}
+
+type githubAuthMiddleware struct{}
+
+func (m *githubAuthMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	githubSecret := []byte(evergreen.GetEnvironment().Settings().Api.GithubWebhookSecret)
+
+	// save the body for future reads
+	payload, err := io.ReadAll(r.Body)
+	r.Body = io.NopCloser(bytes.NewBuffer(payload))
+
+	_, err = github.ValidatePayload(r, githubSecret)
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"source":  "GitHub hook",
+			"message": "rejecting GitHub webhook",
+			"msg_id":  r.Header.Get("X-Github-Delivery"),
+			"event":   r.Header.Get("X-Github-Event"),
+		}))
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(errors.Wrap(err, "validating github payload")))
+		return
+	}
+
+	//reset the body so that it can be parsed again
+	r.Body = io.NopCloser(bytes.NewBuffer(payload))
```

Comment:
Rather than doing this body-resetting logic to try resetting the request's state back to what it was originally, can we instead just set the request context value using `r.WithContext(context.WithValue(r.Context(), ...)`? That way, the route's parse method can just pull the parsed payload out and it'll also avoid the need to re-parse the body a second time.